### PR TITLE
draft kv, doesn't work with auth

### PR DESCRIPTION
**Pull request by levb**

### CODE REVIEW

File: server/api/impl/proxy/proxy.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		up = p.builtIn[app.Manifest.AppID]
+		if up != nil {
+			return up, nil
+		}
+	}
+	if app.Manifest.RemoteRootURL == "" {
+		return nil, errors.New("only built-in and remote http upstreams are supported, hosted AWS Lambda coming soon")
+	}
+
+	// TODO: support AWS Lambda upstream
```
Comment (iomodo) [inline server/api/impl/proxy/proxy.go]: Just to clarify, this is the part where we create AWS client, with appropriate credentials.

File: server/api/impl/proxy/upstream.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		req.Context, err = expander.Expand(sub.Expand)
+		if err != nil {
+			return err
+		}
+
+		// Always set the AppID for routing the request to the App
+		req.Context.AppID = sub.AppID
+
+		go func() {
+			_ = up.InvokeNotification(&req)
```
Comment (iomodo) [inline server/api/impl/proxy/upstream.go]: What is InvokeNotification?

File: server/api/impl/upstream/uphttp/upstream.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package uphttp
```
Comment (iomodo) [inline server/api/impl/upstream/uphttp/upstream.go:4]: This package will be implemented by the aws connector right? Basically my PR has it already implemented(kinda), correct? 
Or are we abstracting away different upstream mechanisms and we will keep this as well as aws one?

File: server/http/restapi/kv.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+func (a *restapi) kvDelete(w http.ResponseWriter, req *http.Request, botUserID, prefix string) {
+	id := mux.Vars(req)["key"]
+	err := a.api.AppServices.KVDelete(botUserID, prefix, id)
+	if err != nil {
+		httputils.WriteInternalServerError(w, err)
+		return
+	}
+}
+
+func (a *restapi) handleKV(
```
Comment (iomodo) [inline server/http/restapi/kv.go:68]: Is this called by the app? How can we sandbox one app KV store from another? One user's data from another? probably JWTs should be checked here? Or can we incorporate OAuth2 scopes here?

File: server/api/impl/proxy/proxy.go (hunk 1 above)
Comment (levb) [reply server/api/impl/proxy/proxy.go]: Right, or return a cached one

File: server/api/impl/proxy/upstream.go (hunk 2 above)
Comment (levb) [reply server/api/impl/proxy/upstream.go]: Notifications (out webhooks) are async, no response expected; with AWS Lambda can and should be invoked as such to minimize the (expensive) concurrency.

File: server/api/impl/upstream/uphttp/upstream.go (hunk 3 above)
Comment (levb) [reply server/api/impl/upstream/uphttp/upstream.go:4]: Abstracting away, we need a new package `upawslambda` that would implement `api.Upstream` for Lambda. We want to keep the `uphttp` practically as is to support remote Apps. I am also keen on the use of "builtIn" for the -apps plugin itself, and perhaps some Apps that we can migrate from go plugins to "built-in" Apps easily.

File: server/http/restapi/kv.go (hunk 4 above)
Comment (levb) [reply server/http/restapi/kv.go:68]: > How can we sandbox one app KV store from another? 

By checking the userID to be that of the Bot, and namespacing accordingly. No user-level isolation inside the app (expect the OAuth2 tokens, coming soon). Similar to current plugins.

Comment (levb) [reply server/http/restapi/kv.go:68]: All KV operations must be performed (for now) with the Bot token. We _could_ partition by the actual userID, but that'd be a bit more complexity for the App to keep track of. I really hope that we figure a better state storage solution soon, but for 1.0 this should be suffiicient?

File: server/api/impl/proxy/upstream.go (hunk 2 above)
Comment (iomodo) [reply server/api/impl/proxy/upstream.go]: Basically lambda functions which do not return info? Not even a success(int terms of function was invoked)?

File: server/api/impl/upstream/uphttp/upstream.go (hunk 3 above)
Comment (iomodo) [reply server/api/impl/upstream/uphttp/upstream.go:4]: "built-in" apps will be same as plugins but with extra layers :)
Understood

File: server/http/restapi/kv.go (hunk 4 above)
Comment (iomodo) [reply server/http/restapi/kv.go:68]: No user partition is fine at this point. But we should maintain data security for customers. What prevents Malicious App to send first customer's userID, get their data and transfer it to the second customer?

File: server/api/impl/proxy/auth.gogo [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package helloapp
+package http_hello
```
Comment (mickmister) [inline server/api/impl/proxy/auth.gogo:1]: Is the file extension `.gogo` intentional?

**Review by mickmister — COMMENTED** [review]
Great foundation :+1: No issues found with the PR

File: server/api/app.go [hunk 6]
```diff
@@ -9,16 +9,31 @@ type Manifest struct {
 	DisplayName string `json:"display_name,omitempty"`
 	Description string `json:"description,omitempty"`
 
-	OAuth2CallbackURL string `json:"oauth2_callback_url,omitempty"`
-	HomepageURL       string `json:"homepage_url,omitempty"`
-	RootURL           string `json:"root_url"`
+	HomepageURL string `json:"homepage_url,omitempty"`
+
+	RootURL                 string `json:"root_url,omitempty"`
+	RemoteOAuth2CallbackURL string `json:"remote_oauth2_callback_url,omitempty"`
```
Comment (levb) [inline server/api/app.go]: no impact to Redux?

File: server/api/app.go [hunk 7]
```diff
@@ -9,16 +9,31 @@ type Manifest struct {
...
 	// "/command/apptrigger"}``.
 	RequestedLocations Locations `json:"requested_locations,omitempty"`
+
+	// By default invoke "/install", expanding App, AdminAccessToken, and Config
+	Install *Call `json:"install,omitempty"`
+
+	// Deployment manifest for hostable apps will include path->invoke mappings
+}
+
+var DefaultInstall = &Call{
```
Comment (levb) [inline server/api/app.go]: check correct defaulting in `InstallApp`

File: server/api/expand.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package api
+
+import "github.com/mattermost/mattermost-server/v5/model"
+
+type ExpandLevel string
+
+// <><> TODO update ExpandLevels in redux
```
Comment (levb) [inline server/api/expand.go]: !!!

File: server/api/impl/admin/install.go [hunk 9]
```diff
@@ -1,23 +1,23 @@
...
 	"github.com/mattermost/mattermost-plugin-apps/server/utils"
 	"github.com/mattermost/mattermost-plugin-apps/server/utils/md"
 	"github.com/mattermost/mattermost-server/v5/model"
 )
 
-func (s *service) InstallApp(cc *apps.Context, sessionToken apps.SessionToken, in *apps.InInstallApp) (*apps.App, md.MD, error) {
-	// TODO check if acting user is a sysadmin
-	app, err := s.GetApp(cc.AppID)
+func (a *Admin) InstallApp(cc *api.Context, sessionToken api.SessionToken, in *api.InInstallApp) (*api.App, md.MD, error) {
+	// TODO <><> check if acting user is a sysadmin
```
Comment (levb) [inline server/api/impl/admin/install.go:19]: // TODO <><> check if acting user is a sysadmin

File: server/api/impl/proxy/bindings.go [hunk 10]
```diff
@@ -43,29 +43,38 @@ func addAppID(bb []*apps.Binding, appID apps.AppID) []*apps.Binding {
...
-		appCC.AppID = app.Manifest.AppID
-		bb, err := s.Client.GetBindings(&appCC)
+		appCC.AppID = appID
+
+		up, err := p.upstreamForApp(app)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to make upstream for %s", appID)
+		}
+
+		// TODO PERF: Add caching
```
Comment (levb) [inline server/api/impl/proxy/bindings.go:48]: ticket these

File: server/api/impl/proxy/expand.go [hunk 11]
```diff
@@ -65,38 +74,41 @@ func (e *expander) Expand(expand *apps.Expand) (*apps.Context, error) {
...
+		user, err := e.mm.User.Get(e.UserID)
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to expand user %s", e.UserID)
 		}
 		e.User = user
 	}
 
-	clone.ExpandedContext = apps.ExpandedContext{
+	// Insert the access tokens
+	// <><>TODO insert user OAuth token if applicable
```
Comment (levb) [inline server/api/impl/proxy/expand.go]: <><>TODO insert user OAuth token if applicable

File: server/api/impl/proxy/upstream.go (hunk 2 above)
Comment (levb) [reply server/api/impl/proxy/upstream.go]: AWS Lambda has "async" mode which is perfectly suited for these.

File: server/api/impl/upstream/upawslambda/upstream.go [hunk 12]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package upawslambda
```
Comment (levb) [inline server/api/impl/upstream/upawslambda/upstream.go:4]: implement

File: server/api/impl/admin/install.go (hunk 9 above)
Comment (levb) [reply server/api/impl/admin/install.go:19]: See if it can be done in mux middleware for /admin?

**Pull request closed without merging**